src/hotspot/share/cds/metaspaceShared.cpp

```diff
@@ -776,12 +780,14 @@ void MetaspaceShared::preload_and_dump_impl(TRAPS) {
   log_info(cds)("Rewriting and linking classes: done");
 
 #if INCLUDE_CDS_JAVA_HEAP
-  if (CDSConfig::is_dumping_heap()) {
-    StringTable::allocate_shared_strings_array(CHECK);
-    ArchiveHeapWriter::init();
-    if (use_full_module_graph()) {
-      HeapShared::reset_archived_object_states(CHECK);
-    }
+  StringTable::allocate_shared_strings_array(CHECK);
+  if (!HeapShared::is_archived_boot_layer_available(THREAD)) {
+    log_info(cds)("archivedBootLayer not available, disabling full module graph");
+    disable_full_module_graph();
+  }
+  ArchiveHeapWriter::init();
+  if (use_full_module_graph()) {
+    HeapShared::reset_archived_object_states(CHECK);
```

Comment:
This block of code needs to be inside `if (CDSConfig::is_dumping_heap())`

Reply:
Thanks for catching this. Fixed.
Also moved the `HeapShared::init_for_dumping(CHECK)` inside the "if" block.